lib/bolt/util.rb

```diff
@@ -169,12 +169,22 @@ def postwalk_vals(data, skip_top = false, &block)
       # references to the same object and prevents endless recursion.
       # Credit to Jan Molic via https://github.com/rubyworks/facets/blob/master/LICENSE.txt
       def deep_clone(obj, cloned = {})
-        cloned[obj.object_id] if cloned.include?(obj.object_id)
+        return cloned[obj.object_id] if cloned.include?(obj.object_id)
+
+        # The Java constant will only be defined when using a JRuby interpreter.
+        # In order to detect JRuby the RUBY_PLATFORM constant is examined instead of
+        # determining if the Java::JavaLang::CloneNotSupportedException constant is defined because
+        # it is possible (for example on the first plan run after a fresh pe-bolt-server install)
+        # that the Java::JavaLang::CloneNotSupportedException is not defined before the exception
```

Comment:
Hmm I'm still not sure. If the class isn't defined until the exception is raised, then we still can't refer to it here. We may need to pull this into a method and call it each time. Then it won't be called until the appropriate `rescue` line is necessary, in which case the exception class will be defined if it's the thing being thrown (and if it's not defined then it can't possibly be what we need to catch anyway).

Reply:
I'm not sure I follow. Are you saying we cant use the constant in the rescue line `rescue TypeError, Java::JavaLang::CloneNotSupportedException` because if *some other* exception is raised we will get an `undefined constant` error? 

I really don't understand *why* this is happening in the first place. I can only get it to reproduce by
1. stop pe-bolt-server
2 uninstall pe-bolt-server
3. re-install pe-bolt-server
4. start pe-bolt server
5. restart pe-orchestration-services
Then only the very first time I run a plan (specifically a plan that uses an inventory method like get_targets) the `Java::JavaLang::CloneNotSupportedException` constant is not defined. This is surprising to me as I would think that restarting both the services would cause it as well.

Reply:
So maybe we would want to go back to one of my earlier attempts:
``ruby
        rescue StandardError => e
          if defined? Java::JavaLang::CloneNotSupportedException
            raise e unless e.is_a?(TypeError) || e.is_a?(Java::JavaLang::CloneNotSupportedException)
          else
            raise e unless e.is_a?(TypeError)
          end
```
That way we would only rescue `Java::JavaLang::CloneNotSupportedException` if it is defined and raise any other error.

Reply:
I'm saying we can't use the constant on the line where we add it to the `error_types` array, if it's not defined at that point.

But it's also true that we can't rely on it existing in the `rescue` line, if we're expecting that the cause of it being defined is something needing to raise it. If a different type of exception is raised instead, then `CloneNotSupportedException` may still not exist at that time.

To me, that indicates that we need to decide which set of exceptions to check for _at the time of the rescue_, which means moving it into a method that's called at the last possible moment (in the `rescue` statement itself).

Reply:
Woah, yeah I have NO idea how adding  `error_types << Java::JavaLang::CloneNotSupportedException if RUBY_PLATFORM == 'java'` works. 

Especially bc with the following: 
```ruby
      def acceptable_clone_error_types
        error_types = [TypeError]
        if defined?(Java::JavaLang::CloneNotSupportedException)
          error_types << Java::JavaLang::CloneNotSupportedException
        end 
        error_types
      end
      # Performs a deep_clone, using an identical copy if the cloned structure contains multiple
      # references to the same object and prevents endless recursion.
      # Credit to Jan Molic via https://github.com/rubyworks/facets/blob/master/LICENSE.txt
      def deep_clone(obj, cloned = {})
        return cloned[obj.object_id] if cloned.include?(obj.object_id)

        begin
          cl = obj.clone
        rescue *acceptable_clone_error_types
          cloned[obj.object_id] = obj
          obj
        else
          cloned[obj.object_id] = cl
          cloned[cl.object_id] = cl

          if cl.is_a? Hash
            obj.each { |k, v| cl[k] = deep_clone(v, cloned) }
          elsif cl.is_a? Array
            cl.collect! { |v| deep_clone(v, cloned) }
          elsif cl.is_a? Struct
            obj.each_pair { |k, v| cl[k] = deep_clone(v, cloned) }
          end

          cl.instance_variables.each do |var|
            v = cl.instance_variable_get(var)
            v_cl = deep_clone(v, cloned)
            cl.instance_variable_set(var, v_cl)
          end

          cl
        end
      end
```
`defined?(Java::JavaLang::CloneNotSupportedException)` is false!

Reply:
Okay so the implementation is solid :thumbsup:

Just need to update the comment to reflect our current understanding